README.md
Font Assets
=============

This little gem helps serve font-face assets in Rails 3.1.  It really does these
two things:

* Responds with "proper" mime types for woff, eot, tff, and svg font files, and
* Sets Access-Control-Allow-Origin response headers for font assets, which Firefox requires for cross domain fonts.

In addition, it will also respond to the pre-flight OPTIONS requests made by
supporting browsers (Firefox).

Install
-------

Add `font_assets` to your Gemfile:

```ruby
gem 'font_assets'
```


Usage
-----

By default, in a Rails application, this gem should Just Work™.  However, the
default settings allow any requesting site to use the linked fonts, due to the
Allowed Origin being '*', by default.  This is only useful for browsers which
support this feature (Firefox), but restricting it to certain domains may be
beneficial.

Set the origin domain that will get set in the `Access-Control-Allow-Origin`
header:

```ruby
# in config/environments/production.rb
config.font_assets.origin = 'http://codeschool.com'
```

The origin domain must match the domain of the site serving the page that is
requesting the font, not the host of the font.  For example, if you are using a
CDN to serve your assets (like CloudFront), and the full path to the font asset
is `http://d3rd6rvl24noqd.cloudfront.net/assets/fonts/Pacifico-webfont-734f1436e605566ae2f1c132905e28b2.woff`,
but the URI the user is visiting is `http://coffeescript.codeschool.com/level/1`,
you'd want to set the origin header to this:

```ruby
config.font_assets.origin = 'http://coffeescript.codeschool.com'
```

An Example Response
-------------------

Below is an example response for a .woff font asset on a Rails 3.1 application
running behind several proxies and caches (including CloudFront):

```
$ curl -i http://d1tijy5l7mg5kk.cloudfront.net/assets/ubuntu/Ubuntu-Bold-webfont-4bcb5239bfd34be67bc07901959fc6e1.woff
HTTP/1.0 200 OK
Server: nginx
Date: Sat, 14 Jan 2012 19:45:19 GMT
Content-Type: application/x-font-woff
Last-Modified: Sat, 14 Jan 2012 16:58:14 GMT
Cache-Control: public, max-age=31557600
Access-Control-Allow-Origin: http://www.codeschool.com
Access-Control-Allow-Methods: GET
Access-Control-Allow-Headers: x-requested-with
Access-Control-Max-Age: 3628800
X-Content-Digest: 66049433125f563329c4178848643536f76459e5
X-Rack-Cache: fresh
Content-Length: 17440
X-Varnish: 311344447
Age: 289983
X-Cache: Hit from cloudfront
X-Amz-Cf-Id: 9yzifs_hIQF_MxPLwSR8zck3eZVXJ8LFKpMUpXnu2SmMuEmyrUbHdQ==,Lbh9kfjr0YRm77seSmOSQ6oFkUEMabvtFStJLhTOy9BfGrIXVneoKQ==
Via: 1.1 varnish, 1.0 2815dd16e8c2a0074b81a6148bd8aa3a.cloudfront.net:11180 (CloudFront), 1.0 f9e7403ca14431787835521769ace98a.cloudfront.net:11180 (CloudFront)
Connection: close
```

In it, you can see where this middleware has injected the `Content-Type` and
`Access-Control-*` headers into the response.

And below is an example OPTIONS request response:

```
$ curl -i -X OPTIONS http://www.codeschool.com/
HTTP/1.1 200 OK
Server: nginx
Date: Wed, 18 Jan 2012 04:13:25 GMT
Connection: keep-alive
Access-Control-Allow-Origin: http://www.codeschool.com
Access-Control-Allow-Methods: GET
Access-Control-Allow-Headers: x-requested-with
Access-Control-Max-Age: 3628800
Vary: Accept-Encoding
X-Rack-Cache: invalidate, pass
Content-Length: 0
```

License
-------

Font Assets is released under the MIT license.

Contributors
------------

* [Nathanial Bibler](https://github.com/nbibler)
* [Eric Allam](https://github.com/rubymaverick)
* [Ryan Montgomery](https://github.com/rmontgomery429)
* [Jack Foy](https://github.com/jfoy)
* [Nick Urban](https://github.com/nickurban)


CHANGELOG.md
### master

### 0.1.11 / 2014-02-25

Bug Fixes

* Handle cases when PATH_INFO is nil or blank. Thanks [arunthampi](https://github.com/arunthampi)!

### 0.1.10 / 2013-07-07

Fixing Regression

### 0.1.9 / 2013-07-07

Bug Fixes

* Change middleware insert order to work with config.threadsafe! (thnx [Nick Urban!](https://github.com/nickurban))

### 0.1.1 / 2012-01-18

[full changelog](https://github.com/rubymaverick/font_assets/compare/v0.1.0...v0.1.1)

Bug Fixes

* Fix Rack::File#empty? call in middleware causing an exception.

### 0.1.0 / 2012-01-18

[full changelog](https://github.com/rubymaverick/font_assets/compare/v0.0.2...v0.1.0)

Enhancements

* Refactor MIME type settings into a local object.
* Add RSpec specs on the middleware and mime type handler.
* Update the README with more detail and examples.
* Fixed a typo in "ttf" in the middleware.
* Removed duplicate Railtie definition.

Bug Fixes

* Fix MIME types not being properly set in recent versions of Rack (starting in [Rack 1.3.0](https://github.com/rack/rack/commit/469518f7d971ba99fc335cf546d605d2364c81aa))

### 0.0.2 / 2011-12-15

[full changelog](https://github.com/rubymaverick/font_assets/compare/v0.0.1...v0.0.2)

Enhancements

* Make MIME setting more future-compatible by optionally setting it.

### 0.0.1 / 2011-12-15

Initial release.


spec/spec_helper.rb
RSpec.configure do |config|
  config.treat_symbols_as_metadata_keys_with_true_values = true
  config.run_all_when_everything_filtered = true
  config.filter_run :focus
end


spec/middleware_spec.rb
require 'spec_helper'
require 'font_assets/middleware'

describe FontAssets::Middleware do
  it 'passes all Rack::Lint checks' do
    app = Rack::Lint.new(load_app)
    request app, '/'
  end

  context 'for GET requests' do
    context 'to font assets' do
      context 'with an app that allows ssl' do
        let(:app) { load_app 'http://test.origin', allow_ssl: true }

        context 'and makes a https request' do
          let(:response) { request app, 'https://test.origin/test.ttf' }

          context 'the response headers' do
            subject { response[1] }

            its(["Access-Control-Allow-Headers"]) { should == "x-requested-with" }
            its(["Access-Control-Max-Age"]) { should == "3628800" }
            its(['Access-Control-Allow-Methods']) { should == 'GET' }
            its(['Access-Control-Allow-Origin']) { should == 'https://test.origin' }
            its(['Content-Type']) { should == 'application/x-font-ttf' }
          end
        end

        context 'and makes a http request' do
          let(:response) { request app, '/test.ttf' }

          context 'the response headers' do
            subject { response[1] }

            its(["Access-Control-Allow-Headers"]) { should == "x-requested-with" }
            its(["Access-Control-Max-Age"]) { should == "3628800" }
            its(['Access-Control-Allow-Methods']) { should == 'GET' }
            its(['Access-Control-Allow-Origin']) { should == 'http://test.origin' }
            its(['Content-Type']) { should == 'application/x-font-ttf' }
          end
        end

      end

      context 'with an app that allows ssl from any origin' do
        let(:app) { load_app '*', allow_ssl: true }

        context 'and makes an https request' do
          let(:response) { request app, 'https://test.origin/test.ttf' }

          context 'the response headers' do
            subject { response[1] }

            its(["Access-Control-Allow-Headers"]) { should == "x-requested-with" }
            its(["Access-Control-Max-Age"]) { should == "3628800" }
            its(['Access-Control-Allow-Methods']) { should == 'GET' }
            its(['Access-Control-Allow-Origin']) { should == '*' }
            its(['Content-Type']) { should == 'application/x-font-ttf' }
          end
        end

        context 'and makes an http request' do
          let(:response) { request app, 'http://test.origin/test.ttf' }

          context 'the response headers' do
            subject { response[1] }

            its(["Access-Control-Allow-Headers"]) { should == "x-requested-with" }
            its(["Access-Control-Max-Age"]) { should == "3628800" }
            its(['Access-Control-Allow-Methods']) { should == 'GET' }
            its(['Access-Control-Allow-Origin']) { should == '*' }
            its(['Content-Type']) { should == 'application/x-font-ttf' }
          end
        end
      end

      context 'with an app with just the origin specified' do
        let(:app) { load_app 'http://test.origin' }
        let(:response) { request app, '/test.ttf' }

        context 'the response headers' do
          subject { response[1] }

          its(["Access-Control-Allow-Headers"]) { should == "x-requested-with" }
          its(["Access-Control-Max-Age"]) { should == "3628800" }
          its(['Access-Control-Allow-Methods']) { should == 'GET' }
          its(['Access-Control-Allow-Origin']) { should == 'http://test.origin' }
          its(['Content-Type']) { should == 'application/x-font-ttf' }
        end
      end
    end

    context 'to non-font assets' do
      let(:app) { load_app }
      let(:response) { request app, '/' }

      context 'the response headers' do
        subject { response[1] }

        its(["Access-Control-Allow-Headers"]) { should be_nil }
        its(["Access-Control-Max-Age"]) { should be_nil }
        its(['Access-Control-Allow-Methods']) { should be_nil }
        its(['Access-Control-Allow-Origin']) { should be_nil }
        its(['Content-Type']) { should == 'text/plain' }
      end
    end

    context 'when PATH_INFO is nil' do
      let(:app) { load_app }
      let(:response) { request app, '/', 'PATH_INFO' => nil }

      context 'the response headers' do
        subject { response[1] }

        its(["Access-Control-Allow-Headers"]) { should be_nil }
        its(["Access-Control-Max-Age"]) { should be_nil }
        its(['Access-Control-Allow-Methods']) { should be_nil }
        its(['Access-Control-Allow-Origin']) { should be_nil }
        its(['Content-Type']) { should == 'text/plain' }
      end
    end

    context 'when PATH_INFO is empty string' do
      let(:app) { load_app }
      let(:response) { request app, '/', 'PATH_INFO' => ''}

      context 'the response headers' do
        subject { response[1] }

        its(["Access-Control-Allow-Headers"]) { should be_nil }
        its(["Access-Control-Max-Age"]) { should be_nil }
        its(['Access-Control-Allow-Methods']) { should be_nil }
        its(['Access-Control-Allow-Origin']) { should be_nil }
        its(['Content-Type']) { should == 'text/plain' }
      end
    end
  end

  context 'for OPTIONS requests' do
    let(:app) { load_app 'http://test.options' }
    let(:response) { request app, '/test.ttf', :method => 'OPTIONS' }

    context 'the response headers' do
      subject { response[1] }

      its(["Access-Control-Allow-Headers"]) { should == "x-requested-with" }
      its(["Access-Control-Max-Age"]) { should == "3628800" }
      its(['Access-Control-Allow-Methods']) { should == 'GET' }
      its(['Access-Control-Allow-Origin']) { should == 'http://test.options' }

      it 'should not contain a Content-Type' do
        subject['Content-Type'].should be_nil
      end
    end

    context 'the response body' do
      subject { response[2] }
      it { should be_empty }
    end
  end


  private


  def load_app(origin = 'http://test.local', options={})
    FontAssets::Middleware.new(inner_app, origin, options)
  end

  def inner_app
    lambda { |env| [200, {'Content-Type' => 'text/plain'}, 'Success'] }
  end

  def request(app, path, options = {})
    app.call Rack::MockRequest.env_for(path, options)
  end
end


spec/mime_types_spec.rb
require 'spec_helper'
require 'font_assets/mime_types'

describe FontAssets::MimeTypes do
  context 'given an empty hash' do
    let(:hash) { Hash.new }
    subject { described_class.new(hash) }

    it 'adds the known mime types' do
      FontAssets::MimeTypes::MIME_TYPES.each_pair do |ext, type|
        subject[ext].should == type
      end
    end
  end

  context 'given a populated hash' do
    let(:default_type) { 'default/type' }
    let(:hash) { { '.ttf' => default_type, '.svg' => 'test/type' } }
    subject { described_class.new(hash, default_type) }

    it 'retains the non-default-matching mime types' do
      subject['.svg'].should == hash['.svg']
    end

    it 'overrides the default-matching mime types' do
      subject['.ttf'].should_not == hash['.ttf']
    end
  end

  context '#[]' do
    let(:types) { described_class.new({}) }

    it 'returns the mime type of the passed extension' do
      types['.woff'].should == 'application/x-font-woff'
    end

    it 'returns the default mime type for unknown extensions' do
      types['.bad'].should == 'application/octet-stream'
    end
  end

  context '#font?' do
    let(:types) { described_class.new({}) }

    it 'is true for known font extensions' do
      FontAssets::MimeTypes::MIME_TYPES.keys.each do |key|
        types.font?(key).should be_true
      end
    end

    it 'is false for unrecognized font extensions' do
      types.font?('.bad').should be_false
    end
  end
end


lib/font_assets/mime_types.rb
module FontAssets
  class MimeTypes
    DEFAULT_TYPE = 'application/octet-stream'
    MIME_TYPES = {
      '.eot' => 'application/vnd.ms-fontobject',
      '.svg' => 'image/svg+xml',
      '.ttf' => 'application/x-font-ttf',
      '.otf' => 'application/x-font-otf',
      '.woff' => 'application/x-font-woff'
    }

    def initialize(types, default = DEFAULT_TYPE.dup)
      @types = types.dup
      @default = default

      MIME_TYPES.each_pair do |extension, type|
        set extension, type
      end
    end

    def [](extension)
      @types.fetch(extension, DEFAULT_TYPE.dup).dup
    end

    def font?(extension)
      MIME_TYPES.keys.include? extension
    end

    def set(extension, mime_type)
      if @types[extension].nil? || @types[extension] == @default
        set!(extension, mime_type)
      end
    end

    def set!(extension, mime_type)
      @types[extension] = mime_type
    end
  end
end


lib/font_assets/middleware.rb
require 'rack'
require 'font_assets/mime_types'

module FontAssets
  class Middleware

    def initialize(app, origin, options={})
      @app = app
      @origin = origin
      @options = options
      @mime_types = FontAssets::MimeTypes.new(Rack::Mime::MIME_TYPES)
    end

    def access_control_headers
      {
        "Access-Control-Allow-Origin" => origin,
        "Access-Control-Allow-Methods" => "GET",
        "Access-Control-Allow-Headers" => "x-requested-with",
        "Access-Control-Max-Age" => "3628800"
      }
    end

    def call(env)
      @ssl_request = Rack::Request.new(env).scheme == "https"
      # intercept the "preflight" request
      if env["REQUEST_METHOD"] == "OPTIONS"
        return [200, access_control_headers, []]
      else
        code, headers, body = @app.call(env)
        set_headers! headers, body, env["PATH_INFO"]
        [code, headers, body]
      end
    end


    private

    def origin
      if !wildcard_origin? and allow_ssl? and ssl_request?
        uri = URI(@origin)
        uri.scheme = "https"
        uri.to_s
      else
        @origin
      end
    end

    def wildcard_origin?
      @origin == '*'
    end

    def ssl_request?
      @ssl_request
    end

    def allow_ssl?
      @options[:allow_ssl]
    end

    def extension(path)
      if path.nil? || path.length == 0
        nil
      else
        "." + path.split("?").first.split(".").last
      end
    end

    def font_asset?(path)
      @mime_types.font? extension(path)
    end

    def set_headers!(headers, body, path)
      if ext = extension(path) and font_asset?(ext)
        headers.merge!(access_control_headers)
        headers.merge!('Content-Type' => mime_type(ext)) if headers['Content-Type']
      end
    end

    def mime_type(extension)
      @mime_types[extension]
    end
  end
end
